### ISSUE TITLE
Map new WoS record format(s) to pub_hash

### ISSUE THREAD

**Issue by peetucket** [sul-dlss/sul_pub#274]
prereq #184

- [ ] confirm that the new WoS records have the minimum required fields to build pub_hash citations; examine a record that exists in both the Sciencewire and Web of Science APIs and compare citations that would be created from both source records and see how they compare

DLW:  A key piece of that puzzle is to determine what the minimal set of fields is to create a citation.  This involves review of
- https://github.com/sul-dlss/sul_pub/blob/master/app/models/pub_hash.rb#L56
- because that is where the pub_hash fields are extracted to create a citation

**Comment by Maatary** [comment]
A first glimpse at the actual first step of the mapping from science wire to a pub_hash (this is not the complete set of information but the bulk)

in the ScienceWireHarvester

```
pub_hash = SciencewireSourceRecord.convert_sw_publication_doc_to_hash(incoming_sw_xml_doc)
```


in SciencewireSourceRecord


```
def self.convert_sw_publication_doc_to_hash(publication)
    doi = extract_doi(publication)
    issn = extract_issn(publication)
    pmid = extract_pmid(publication)
    swid = extract_swid(publication)
    wosid = extract_wosid(publication)

    doi_identifier = { type: 'doi', id: doi, url: "#{Settings.DOI.BASE_URI}#{doi}" } unless doi.blank?
    issn_identifier = { type: 'issn', id: issn, url: Settings.SULPUB_ID.SEARCHWORKS_URI + issn } unless issn.blank?

    identifiers = []
    identifiers << { type: 'PMID', id: pmid, url: "#{Settings.PUBMED.ARTICLE_BASE_URI}#{pmid}" } unless pmid.blank?
    identifiers << { type: 'WoSItemID', id: wosid, url: "#{Settings.SCIENCEWIRE.ARTICLE_BASE_URI}#{wosid}" } unless wosid.blank?
    identifiers << { type: 'PublicationItemID', id: swid } unless swid.blank?
    identifiers << doi_identifier unless doi.blank?

    record_as_hash = {}
    record_as_hash[:provenance] = Settings.sciencewire_source
    record_as_hash[:sw_id] = swid
    record_as_hash[:pmid] = pmid unless pmid.blank?
    record_as_hash[:issn] = issn unless issn.blank?
    record_as_hash[:identifier] = identifiers

    record_as_hash[:title] = publication.xpath('Title').text unless publication.xpath('Title').blank?
    record_as_hash[:abstract_restricted] = publication.xpath('Abstract').text unless publication.xpath('Abstract').blank?
    record_as_hash[:author] = publication.xpath('AuthorList').text.split('|').collect { |author| { name: author } }

    record_as_hash[:year] = publication.xpath('PublicationYear').text unless publication.xpath('PublicationYear').blank?
    record_as_hash[:date] = publication.xpath('PublicationDate').text unless publication.xpath('PublicationDate').blank?

    record_as_hash[:authorcount] = publication.xpath('AuthorCount').text unless publication.xpath('AuthorCount').blank?

    record_as_hash[:keywords_sw] = publication.xpath('KeywordList').text.split('|') unless publication.xpath('KeywordList').blank?
    record_as_hash[:documenttypes_sw] = publication.xpath('DocumentTypeList').text.split('|')

    record_as_hash[:documentcategory_sw] = publication.xpath('DocumentCategory').text unless publication.xpath('DocumentCategory').blank?
    sul_document_type = lookup_cap_doc_type_by_sw_doc_category(record_as_hash[:documentcategory_sw])
    record_as_hash[:type] = sul_document_type

    record_as_hash[:publicationimpactfactorlist_sw] = publication.xpath('PublicationImpactFactorList').text.split('|') unless publication.xpath('PublicationImpactFactorList').blank?
    record_as_hash[:publicationcategoryrankinglist_sw] = publication.xpath('PublicationCategoryRankingList').text.split('|') unless publication.xpath('PublicationCategoryRankingList').blank?
    record_as_hash[:numberofreferences_sw] = publication.xpath('NumberOfReferences').text unless publication.xpath('NumberOfReferences').blank?
    record_as_hash[:timescited_sw_retricted] = publication.xpath('TimesCited').text unless publication.xpath('TimesCited').blank?
    record_as_hash[:timenotselfcited_sw] = publication.xpath('TimesNotSelfCited').text unless publication.xpath('TimesNotSelfCited').blank?
    record_as_hash[:authorcitationcountlist_sw] = publication.xpath('AuthorCitationCountList').text unless publication.xpath('AuthorCitationCountList').blank?
    record_as_hash[:rank_sw] = publication.xpath('Rank').text unless publication.xpath('Rank').blank?
    record_as_hash[:ordinalrank_sw] = publication.xpath('OrdinalRank').text unless publication.xpath('OrdinalRank').blank?
    record_as_hash[:normalizedrank_sw] = publication.xpath('NormalizedRank').text unless publication.xpath('NormalizedRank').blank?
    record_as_hash[:newpublicationid_sw] = publication.xpath('NewPublicationItemID').text unless publication.xpath('NewPublicationItemID').blank?
    record_as_hash[:isobsolete_sw] = publication.xpath('IsObsolete').text unless publication.xpath('IsObsolete').blank?

    record_as_hash[:publisher] = publication.xpath('CopyrightPublisher').text unless publication.xpath('CopyrightPublisher').blank?
    record_as_hash[:city] = publication.xpath('CopyrightCity').text unless publication.xpath('CopyrightCity').blank?
    record_as_hash[:stateprovince] = publication.xpath('CopyrightStateProvince').text unless publication.xpath('CopyrightStateProvince').blank?
    record_as_hash[:country] = publication.xpath('CopyrightCountry').text unless publication.xpath('CopyrightCountry').blank?
    record_as_hash[:pages] = publication.xpath('Pagination').text unless publication.xpath('Pagination').blank?

    if sul_document_type == Settings.sul_doc_types.inproceedings
      conference_hash = {}
      conference_hash[:name] = publication.xpath('ConferenceTitle').text unless publication.xpath('ConferenceTitle').blank?
      conference_hash[:startdate] = publication.xpath('ConferenceStartDate').text unless publication.xpath('ConferenceStartDate').blank?
      conference_hash[:enddate] = publication.xpath('ConferenceEndDate').text unless publication.xpath('ConferenceEndDate').blank?
      conference_hash[:city] = publication.xpath('ConferenceCity').text unless publication.xpath('ConferenceCity').blank?
      conference_hash[:statecountry] = publication.xpath('ConferenceStateCountry').text unless publication.xpath('ConferenceStateCountry').blank?
      record_as_hash[:conference] = conference_hash unless conference_hash.empty?
    elsif sul_document_type == Settings.sul_doc_types.book
      record_as_hash[:booktitle] = publication.xpath('PublicationSourceTitle').text unless publication.xpath('PublicationSourceTitle').blank?
      record_as_hash[:pages] = publication.xpath('Pagination').text unless publication.xpath('Pagination').blank?
    end

    if sul_document_type == Settings.sul_doc_types.article || (sul_document_type == Settings.sul_doc_types.inproceedings && !publication.xpath('Issue').blank?)
      journal_hash = {}
      journal_hash[:name] = publication.xpath('PublicationSourceTitle').text unless publication.xpath('PublicationSourceTitle').blank?
      journal_hash[:volume] = publication.xpath('Volume').text unless publication.xpath('Volume').blank?
      journal_hash[:issue] = publication.xpath('Issue').text unless publication.xpath('Issue').blank?
      journal_hash[:articlenumber] = publication.xpath('ArticleNumber').text unless publication.xpath('ArticleNumber').blank?
      journal_hash[:pages] = publication.xpath('Pagination').text unless publication.xpath('Pagination').blank?
      journal_identifiers = []
      journal_identifiers << issn_identifier unless issn.blank?
      journal_identifiers << doi_identifier unless doi.blank?
      journal_hash[:identifier] = journal_identifiers
      record_as_hash[:journal] = journal_hash
    end

    # unless issn.blank? || publication.xpath('Issue').blank? || sul_document_type == Settings.sul_doc_types.article
    #   book_series_hash = {}
    #   book_series_hash[:identifier] = [issn_identifier]
    #   book_series_hash[:title] = publication.xpath('PublicationSourceTitle').text unless publication.xpath('PublicationSourceTitle').blank?
    #   book_series_hash[:volume] = publication.xpath('Volume').text unless publication.xpath('Volume').blank?
    #   record_as_hash[:series] = book_series_hash
    # end
    record_as_hash
  end
```

**Comment by Maatary** [comment]
So far, as far as my analysis, the only thing missing at that level are the pubmed metadata. 

```
:mesh_headings
:abstract
[:identifier] << pmc_id 
```

**Comment by Maatary** [comment]
Next step: 

1 - Can we get all of these information from WOS record (Wos or Medline). 

2 - How do we get what is missing from wos (pubmed metadata)

**Comment by Maatary** [comment]
Create a to_ash for an XML coming back from a Web of Science Record

**Comment by Maatary** [comment]
Working on the pure mapping that is making sure that everything that was extracted from a ScienceWire Record can be extracted today (even if the terminology change or not or position and so on). I can say as of now that the most important things are there. However to be exhaustive about all the fields it will require some digging. I need to find the publication that have some of the field that are optional and not always there, and see that in the new version those field are there or not. 

Bottom line is: given that some of the field are optional, i can't assert if the field has been removed or not. So i need records that are supposed to have those optional field and check if they are present in the WOS API. 

Else the other solution, is to explicitly ask them for what they have done with some specific field. 

I believe we could do both in parallel.

**Comment by darrenleeweber** [comment]
Focus only on the fields required for citations, see
- https://github.com/sul-dlss/sul_pub/blob/master/app/models/pub_hash.rb#L56

**Comment by Maatary** [comment]
Mapping available at https://github.com/sul-dlss/sul_pub/wiki/WOS-to-Pub_hash-Mapping
